cilium/CFP-31752-active-connection-tracking.md

```diff
@@ -0,0 +1,200 @@
+Aware Routing - fill the local zone to capacity and then spillover to other
+zones. 
+
+One of the most useful metrics for determining how "full" an endpoint is would
+be how many active and new connections an endpoint is handling. This information
+can be paired with overall resource utilization information for each endpoint to
+make more effective routing decisions.
+
+Although this use case initially surfaced in the xDS discussion, we believe it's
+useful beyond just xDS integration. If we can track the number of outgoing
+connections per zone and Service, this data will likely be useful far beyond
+xDS. 
+
+We propose adding this feature independently from the xDS proposal, and reusing
+it for xDS only if/when that broader CFP is accepted.
+
+
+## Goals
+
+* Add metrics to make it easier to observe traffic distribution across zones
+* Surface information needed by control plane to better distribute traffic
+
+
+## Non-Goals
+
+* Change the behavior of load balancers
+* Supplement or replace data provided by Hubble
+
+
+## Proposal 1: Introduce a new BPF map: cilium_lb_lrs
+
+### Overview
+
+Name: `cilium_lb_lrs`
+
+Key: (`svc_id` - 2 bytes, `zone` - 1 byte, `pad` - 1 byte)
+
+Value: (`opened` - 4 bytes, `closed` - 4 bytes)
+
+When connection to a service is opened
+[[ct\_create4](https://github.com/cilium/cilium/blob/v1.15.3/bpf/lib/conntrack.h#L980),
+[ct\_create6](https://github.com/cilium/cilium/blob/v1.15.3/bpf/lib/conntrack.h#L907)]
+we will fetch zone ID from the selected backend. Then we will fetch the entry
+and increment `opened` counter.
+
+When connection to a service is closed
+[[\_\_ct\_lookup](https://github.com/cilium/cilium/blob/v1.15.3/bpf/lib/conntrack.h#L312)]
+we will use `rev_nat_id` (which corresponds to `svc_id`) and `backend_id` (which
+is stored in `rx_bytes` for `CT_SERVICE` entries). We will resolve zone ID by
+fetching the corresponding backend entry and then increment the `closed`
+counter.
+
+Unfortunately it doesn't catch all cases. One example is "connection refused".
+These entries are removed by GC process in the agent. Active connection tracking
+process should combine data from both sources (GC and BPF map) in order to
+provide accurate gauges for new and ongoing connections.
+
+Alternative solution would count only the connections that have already seen a
+reply SYN packet.However, in this approach we won't be able to extend this
+work to UDP.
+
+### Pros
+
+* Fast: additional code runs only on connection open and close
+* `uint32` for counters guarantees atomic operations
+
+### Cons
+
+* Some events may be missed by datapath so the information must be combined with
+  Agent's (i.e. counting service entries that were GCed)
+* `uint32` will overflow, so metrics reader will need to adjust for that
+
+### Alternatives
+
+* Use event-based mechanism (for example: trace monitoring / Hubble) -
+  unreliable, too slow
+* Dump conntrack table - too slow
```

Comment:
Have you considered batch ops? I assume the "too slow" comment here is because it is slow to dump the map element by element. 

The CT BPF map pressure metric uses batch ops to dump the map and count the elements for the metric already.

Reply:
This looks promising. I need to explore it further:
1. In the initial testing it shows:
    ```
    48791 entries -> 36.65ms
    48803 entries -> 36.38ms
    ```
    which gives ~0.7µs per entry which completely satisfies the scalability requirement.
2. Does this warning apply? https://github.com/cilium/ebpf/blob/main/map.go#L980-L982
3. Are there any other downsides (i.e. am I guaranteed to get the full state? do I get only current/valid entries?)